<a href="https://colab.research.google.com/github/bsmrvl/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module2-make-features/LS_DS_112_Make_Features.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 2*

---

# Make Features 

- Student should be able to understand the purpose of feature engineering
- Student should be able to work with strings in pandas
- Student should be able to work with dates and times in pandas
- Student should be able to modify or create columns of a dataframe using the `.apply()` function


Helpful Links:
- [Minimally Sufficient Pandas](https://medium.com/dunder-data/minimally-sufficient-pandas-a8e67f2a2428)
- [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)
- Python Data Science Handbook
  - [Chapter 3.10](https://jakevdp.github.io/PythonDataScienceHandbook/03.10-working-with-strings.html), Vectorized String Operations
  - [Chapter 3.11](https://jakevdp.github.io/PythonDataScienceHandbook/03.11-working-with-time-series.html), Working with Time Series
- [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)

# [Objective](#feature-engineering) - The Purpose of Feature Engineering



## Overview

Feature Engineering is the process of using a combination of domain knowledge, creativity and the pre-existing columns of a dataset to create completely new columns.

 Machine Learning models try to detect patterns in the data and then associate those patterns with certain predictions. The hope is that by creating new columns on our dataset that we can expose our model to new patterns in the data so that it can make better and better predictions.

This is largely a matter of understanding how to work with individual columns of a dataframe with Pandas --which is what we'll be practicing today!

## Follow Along

Columns of a dataframe hold each hold a specific type of data. Lets inspect some of the common datatypes found in datasets and then we'll make a new feature on a dataset using pre-existing columns.

In [ ]:
import pandas as pd

# Pandas Display Options:
pd.set_option('display.max_rows', 150)
pd.set_option('display.max_columns', 100)

In [ ]:
# Lets take a look at the Ames Iowa Housing Dataset:
df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')

### Specific Columns hold specific kinds of data

In [ ]:
!curl https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv

In [ ]:
df.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
Alley             object
LotShape          object
LandContour       object
Utilities         object
LotConfig         object
LandSlope         object
Neighborhood      object
Condition1        object
Condition2        object
BldgType          object
HouseStyle        object
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle         object
RoofMatl          object
Exterior1st       object
Exterior2nd       object
MasVnrType        object
MasVnrArea       float64
ExterQual         object
ExterCond         object
Foundation        object
BsmtQual          object
BsmtCond          object
BsmtExposure      object
BsmtFinType1      object
BsmtFinSF1         int64
BsmtFinType2      object
BsmtFinSF2         int64
BsmtUnfSF          int64
TotalBsmtSF        int64
Heating           object


In [ ]:
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd       None   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   
1       1262             0             1         2         0             3   
2       1786             1             0         2         1             3   
3       1717             1             0         1         0             3   
4       2198             1             0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd            

In [ ]:
df['BedroomAbvGr'].dtypes

dtype('int64')

In [ ]:
df['LotFrontage'].value_counts(dropna=False)

NaN      259
60.0     143
70.0      70
80.0      69
50.0      57
75.0      53
65.0      44
85.0      40
78.0      25
21.0      23
90.0      23
68.0      19
24.0      19
64.0      19
73.0      18
72.0      17
79.0      17
63.0      17
55.0      17
100.0     16
51.0      15
66.0      15
74.0      15
52.0      14
59.0      13
71.0      12
67.0      12
57.0      12
82.0      12
43.0      12
40.0      12
76.0      11
69.0      11
53.0      10
92.0      10
88.0      10
34.0      10
86.0      10
77.0       9
35.0       9
44.0       9
84.0       9
62.0       9
93.0       8
61.0       8
98.0       8
96.0       8
107.0      7
95.0       7
120.0      7
58.0       7
41.0       6
105.0      6
54.0       6
94.0       6
48.0       6
110.0      6
30.0       6
89.0       6
81.0       6
91.0       6
36.0       6
47.0       5
83.0       5
37.0       5
87.0       5
32.0       5
56.0       5
102.0      4
42.0       4
49.0       4
99.0       3
45.0       3
104.0      3
130.0      3
103.0      3
108.0      3

Some columns hold integer values like the `BedroomAbvGr` which stands for "Bedrooms Above Grade." This is the number of non-basement bedrooms in the home.

For more information on specific column meanings view the [data dictionary](https://github.com/ryanleeallred/datasets/blob/master/Ames%20Housing%20Data/data_description.txt).

In [ ]:
# Look at the first ten rows of the `BedroomAbvGr` column.
# Looks like integers to me!


Some columns hold float values like the `LotFrontage` column.

In [ ]:
# Look at the first ten rows of the `BedroomAbvGr` column.


Hmmm, do the values above look like floats to you?

They all have .0 on them so technically they're being stored as floats, but *should* they be stored as floats?

Lets see what all of the possible values for this column are.

Looks to me like the `LotFrontage` column originally held integer values but was cast to a `float` meaning that each original integer values was converted to its corresponding float representation. 

Any guesses as to why that would have happened?


HINT: What's the most common `LotFrontage` value for this column?

In [ ]:
# NaN is the most common value in this column. What is a NaN

import numpy as np
type(np.NaN)

float

`NaN` stands stands for "Not a Number" and is the default missing value indicator with Pandas. This means there were cells in this column that didn't have a LotFrontage value recorded for those homes. 

This is where domain knowledge starts to come in. Think about the context we're working with here: houses. What might a null or blank cell representing "Linear feet of street connected to property" mean in the context of a housing dataset?

Ok, so maybe it makes seanse to have some NaNs in this column. What is the datatype of a NaN value?

Perhaps some of this data is truly missing or unrecorded data, but sometimes `NaNs` are more likely to indicate something that was "NA" or "Not Applicable" to a particular observation. There could be multiple reasons why there was no value recorded for a particular feature.

Remember, that Pandas tries to maintain a single datatype for all values in a column, and therefore...

In [ ]:
import numpy as np

# What is the datatype of NaN?

The datatype of a NaN is float!  This means that if we have a column of integer values, but the column has even a single `NaN` that column will not be treated with the integer datatype but all of the integers will be converted to floats in order to try and preserve the same datatype throughout the entire column.

You can see already how understanding column datatypes is crucial to understanding how Pandas help us manage our data.

### Making new Features

Lets slim down the dataset and consider just a few specific columns:

- `TotalBsmtSF`
- `1stFlrSF`
- `2ndFlrSF`
- `SalePrice1`


In [ ]:
# I can make a smaller dataframe with a few specific column headers
# by passing a list of column headers inside of the square brackets
df_slim = df[['TotalBsmtSF', '1stFlrSF','2ndFlrSF','SalePrice']].copy()


### Syntax for creating new columns

When making a new column on a dataframe, we have to use the square bracket syntax of accessing a column. We can't use "dot syntax" here.

In [ ]:
# Lets add up all of the square footage to get a single square footage 
# column for the entire dataset
df_slim['TotalSquareFootage'] = df_slim['TotalBsmtSF'] + df_slim['1stFlrSF'] + df_slim['2ndFlrSF']
df_slim.head()


# Using bracket syntax to make a new 'TotalSquareFootage' column

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage
0          856       856       854     208500                2566
1         1262      1262         0     181500                2524
2          920       920       866     223500                2706
3          756       961       756     140000                2473
4         1145      1145      1053     250000                3343

In [ ]:
# Lets make a nother new column that is 'PricePerSqFt' by
# dividing the price by the square footage

df_slim['PricePerSqFt'] = df_slim['SalePrice'] / df_slim['TotalSquareFootage']
df_slim.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage  \
0          856       856       854     208500                2566   
1         1262      1262         0     181500                2524   
2          920       920       866     223500                2706   
3          756       961       756     140000                2473   
4         1145      1145      1053     250000                3343   

   PricePerSqFt  
0     81.254871  
1     71.909667  
2     82.594235  
3     56.611403  
4     74.783129

Ok, we have made two new columns on our small dataset.

- What does a **high** `PricePerSqFt` say about a home that the square footage and price alone don't capture as directly?

- What does a **low** `PricePerSqFt` say about a home that the square footage and price alone don't directly capture?



## Challenge

I hope you can see how we have used existing columns to create a new column on a dataset that say something new about our unit of observation. This is what making new features (columns) on a dataset is all about and why it's so essential to data science --particularly predictive modeling "Machine Learning." 

We'll spend the rest of the lecture and assignment today trying to get as good as we can at manipulating (cleaning) and creating new columns on datasets.

# [Objective](#work-with-strings) Work with Strings with Pandas

## Overview

So far we have worked with numeric datatypes (ints and floats) but we haven't worked with any columns containing string values. We can't simply use arithmetic to manipulate string values, so we'll need to learn some more techniques in order to work with this datatype.

## Follow Along

We're going to import a new dataset here to work with. This dataset is from LendingClub and holds information about loans issued in Q4 of 2018. This dataset is a bit messy so it will give us plenty of opportunities to clean up existing columns as well as create new ones.

The `!wget` shell command being used here does exactly the same thing that your browser does when you type a URL in the address. It makes a request or "gets" the file at that address. However, in our case the file isn't a webpage, it's a compressed CSV file. 

Try copying and pasting the URL from below into your browser, did it start an automatic download? Any URLs like this that start automatic downloads when navigated to can be used along with the `!wget` command to bring files directly into your notebook's memory.

### Load a new dataset

In [ ]:
!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip



--2020-09-01 17:00:50--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 54.148.13.215, 35.161.89.82, 54.244.115.45
Connecting to resources.lendingclub.com (resources.lendingclub.com)|54.148.13.215|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip’

LoanStats_2018Q4.cs     [               <=>  ]  22.28M  6.70MB/s    in 4.0s    

2020-09-01 17:00:54 (5.56 MB/s) - ‘LoanStats_2018Q4.csv.zip’ saved [23360898]



We need to use the `!unzip` command to extract the csv from the zipped folder.

In [ ]:
!unzip LoanStats_2018Q4.csv.zip

Archive:  /content/LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


We can also use bash/shell commands to look at the raw file using the `!head` and `!tail` commands

In [ ]:
!head /content/LoanStats_2018Q4.csv

As we look at the raw file itself, do you see anything that might cause us trouble as we read in the CSV file to a dataframe?

In [ ]:
# Read in the CSV

lstats = pd.read_csv('LoanStats_2018Q4.csv')

In [ ]:
lstats.tail()

The extra rows at the top and bottom of the file have done two things:

1) The top row has made it so that the entire dataset is being interpreted as column headers

2) The bottom two rows have been read into the 'id' column and are causing every column to have at least two `NaN` values in it.

In [ ]:
# We can fix the header problem by using the 'skiprows' parameter

lstats = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1)
lstats.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  NaN        NaN    27975.0      27975.0          27975.0   36 months   
1  NaN        NaN    10000.0      10000.0          10000.0   60 months   
2  NaN        NaN    28000.0      28000.0          28000.0   60 months   
3  NaN        NaN    16000.0      16000.0          16000.0   60 months   
4  NaN        NaN     7500.0       7500.0           7500.0   36 months   

  int_rate  installment grade sub_grade        emp_title emp_length  \
0   14.47%       962.52     C        C2       Conductor   10+ years   
1   12.98%       227.43     B        B5          Printer    9 years   
2   13.56%       645.15     C        C1  Project Manager  10+ years   
3   13.56%       368.66     C        C1             LMSW  10+ years   
4   10.72%       244.55     B        B2  Program Manager    2 years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0       MORTGAGE    180000.0        Not Verified  Dec-2018   Fully Paid   
1           RENT     60000.0        Not Verified  Dec-2018      Current   
2       MORTGAGE    128500.0     Source Verified  Dec-2018      Current   
3           RENT     46000.0        Not Verified  Dec-2018      Current   
4           RENT     84000.0        Not Verified  Dec-2018  Charged Off   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN         credit_card  Credit card refinancing    117xx   
1          n  NaN   NaN         credit_card  Credit card refinancing    660xx   
2          n  NaN   NaN         credit_card  Credit card refinancing    760xx   
3          n  NaN   NaN         credit_card  Credit card refinancing    787xx   
4          n  NaN   NaN  debt_consolidation       Debt consolidation    600xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  11.47          0.0         Jul-1995             0.0   
1         KS  14.90          0.0         May-2007             0.0   
2         TX  27.35          1.0         Jul-1987             0.0   
3         TX  11.09          1.0         Oct-1997             1.0   
4         IL   3.86          0.0         May-2005             1.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                    39.0                     NaN      10.0      0.0   
1                     NaN                   112.0       7.0      1.0   
2                    24.0                     NaN      16.0      0.0   
3                    11.0                     NaN       9.0      0.0   
4                     NaN                   114.0       9.0      1.0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0    29711.0      66.8%       19.0                   w       0.00   
1    10677.0      54.2%       12.0                   w    7352.95   
2    55206.0        75%       31.0                   w   20667.91   
3    18946.0      32.1%       21.0                   w   11503.84   
4     2200.0      12.6%       17.0                   w       0.00   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00  31804.529849         31804.53         27975.00   
1        7352.95   4487.310000          4487.31          2647.05   
2       20667.91  12871.360000         12871.36          7332.09   
3       11503.84   7570.880000          7570.88          4496.16   
4           0.00   2829.930000          2829.93          1465.61   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0        3829.53                 0.0        0.00                   0.0000   
1        1840.26                 0.0        0.00                   0.0000   
2        5539.27                 0.0        0.00                   0.0000   
3        3074.72                 0.0        0.00                   0.0000   
4         484.09                 0.0      880.23                 158.4414   

  last_pymnt_d  last_pymnt_amnt next_pymnt_d last_credit_pu

In [ ]:
lstats.tail()

Lets look at the NaN values of each column so that you can see the problem that the extra rows at the bottom of the file are creating for us

In [ ]:
# Sum null values by column and sort from least to greatest

lstats.isnull().sum().sort_values()

In [ ]:
# Address the extra NaNs in each column by skipping the footer as well.

lstats = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1, skipfooter=2)
lstats.tail()

In [ ]:
print(lstats.shape)
lstats.isnull().sum().sort_values()

(128412, 144)


inq_fi                                             0
mo_sin_old_rev_tl_op                               0
delinq_amnt                                        0
chargeoff_within_12_mths                           0
acc_open_past_24mths                               0
inq_last_12m                                       0
total_cu_tl                                        0
total_rev_hi_lim                                   0
open_rv_24m                                        0
open_rv_12m                                        0
total_bal_il                                       0
open_il_24m                                        0
open_il_12m                                        0
open_act_il                                        0
open_acc_6m                                        0
tot_cur_bal                                        0
tot_coll_amt                                       0
acc_now_delinq                                     0
application_type                              

For good measure, we'll also drop some columns that are made up completely of NaN values.

Why might LendingClub have included columns in their dataset that are 100% blank?

In [ ]:
lstats.drop(['id', 'member_id','desc','url'], axis=1, inplace=True)
lstats.isnull().sum().sort_values()

loan_amnt                                          0
mo_sin_rcnt_rev_tl_op                              0
mo_sin_old_rev_tl_op                               0
delinq_amnt                                        0
chargeoff_within_12_mths                           0
acc_open_past_24mths                               0
total_cu_tl                                        0
inq_fi                                             0
total_rev_hi_lim                                   0
max_bal_bc                                         0
open_rv_24m                                        0
open_rv_12m                                        0
total_bal_il                                       0
open_il_24m                                        0
open_il_12m                                        0
open_act_il                                        0
open_acc_6m                                        0
tot_cur_bal                                        0
tot_coll_amt                                  

### Clean up the `int_rate` column

When we're preparing a dataset for a machine learning model we typically want to represent don't want to leave any string values in our dataset --because it's hard to do math on words. 

Specifically, we have a column that is representing a numeric value, but currently doesn't have a numeric datatype. Lets look at the first 10 values of the `int_rate` column

In [ ]:
# Look at the first 10 values of the int_rate column
lstats['int_rate'].head(10)

0     14.47%
1     12.98%
2     13.56%
3     13.56%
4     10.72%
5     20.89%
6     26.31%
7     23.40%
8     19.92%
9     17.97%
Name: int_rate, dtype: object

In [ ]:
# Look at a specific value from the int_rate column
one = lstats['int_rate'][3]

Problems that we need to address with this column:

- String column that should be numeric
- Percent Sign `%` included with the number
- Leading space at the beginning of the string

However, we're not going to try and write exactly the right code to fix this column in one go. We're going to methodically build up to the code that will help us address these problems.


In [ ]:
# Lets start with just fixing a single string.
# If we can fix one, we can usually fix all of them

In [ ]:
float(one)

ValueError: ignored

In [ ]:
one[1:-1]

'13.56'

In [ ]:
float(one[1:-1])

13.56

In [ ]:
# "Cast" the string value to a float

### Write a function to make our solution reusable!

In [ ]:
# Write a function that can do what we have written above to any 
# string that is passsed to it.

def int_rate_to_float(inst):
  return float(inst[1:-1])

In [ ]:
# Test out our function by calling it on our example
int_rate_to_float(lstats['int_rate'][4])

10.72

In [ ]:
# is the data type correct?
type(int_rate_to_float(lstats['int_rate'][4]))

float

### Apply our solution to every cell in a column

In [ ]:
lstats['int_rate'] = lstats['int_rate'].apply(int_rate_to_float)

In [ ]:
# What type of data is held in our new column?

lstats['int_rate'][5]
# Look at the datatypes of the last 5 columns

20.89

In [ ]:
lstats['int_rate'].mean()

12.928038734699749

## Challenge

We can create a new column with our cleaned values or overwrite the original, whatever we think best suits our needs. On your assignment you will take the same approach in trying to methodically build up the complexity of your code until you have a few lines that will work for any cell in a column. At that point you'll contain all of that functionality in a reusable function block and then use the `.apply()` function to... well... apply those changes to an entire column.

# [Objective](#pandas-apply) Modify and Create Columns using `.apply()`



## Overview

We're already seen one example of using the `.apply()` function to clean up a column. Lets see if we can do it again, but this time on a slightly more complicated use case.

Remember, the goal here is to write a function that will work correctly on any **individual** cell of a specific column. Then we can reuse that function on those individual cells of a dataframe column via the `.apply()` function.

Lets clean up the emp_title "Employment Title" column!

## Follow Along

First we'll try and diagnose how bad the problem is and what improvements we might be able to make.

In [ ]:
# Look at the top 20 employment titles
lstats['emp_title'].value_counts().head(20)

Teacher                     2090
Manager                     1773
Registered Nurse             952
Driver                       924
RN                           726
Supervisor                   697
Sales                        580
Project Manager              526
General Manager              523
Office Manager               521
Owner                        420
Director                     402
Truck Driver                 387
Operations Manager           387
Nurse                        326
Engineer                     325
Sales Manager                304
manager                      301
Supervisor                   270
Administrative Assistant     269
Name: emp_title, dtype: int64

In [ ]:
# How many different unique employment titles are there currently?
lstats['emp_title'].nunique()

43892

In [ ]:
# How often is the employment_title null?

What are some possible reasons as to why a person's employment title may have not been provided?

In [ ]:
# Create some examples that represent the cases that we want to clean up
import numpy as np
examples = ['roger', ' Undue', np.NaN, 'OWNER']

In [ ]:
# Write a function to clean up these use cases and increase uniformity.
def title_clean(inst):
  if(type(inst) == float):
    return 'Unknown'
  else:
    return inst.strip().title()

In [ ]:
# list comprehensions can combine function calls and for loops over lists
# into one succinct and fairly readable single line of code.

for inst in examples:
  print(title_clean(inst))

Roger
Undue
Unknown
Owner


In [ ]:
# We have a function that works as expected. Lets apply it to our column.
# This time we'll overwrite the original column

lstats['emp_title'] = lstats['emp_title'].apply(title_clean)
lstats['emp_title']

0                    Conductor
1                      Printer
2              Project Manager
3                         Lmsw
4              Program Manager
                  ...         
128407          Tax Consultant
128408          Security Guard
128409            Payoff Clerk
128410                 Unknown
128411    Key Accounts Manager
Name: emp_title, Length: 128412, dtype: object

We can use the same code as we did earlier to see how much progress was made.


In [ ]:
# Look at the top 20 employment titles

In [ ]:
# How many different unique employment titles are there currently?
lstats['emp_title'].nunique()

34902

In [ ]:
# How often is the employment_title null (NaN)?

## Challenge

Using the .apply() function isn't always about creating new columns on a dataframe, we can use it to clean up or modify existing columns as well. 

# [Objective](#dates-and-times) Work with Dates and Times with Pandas

## Overview

Pandas has its own datatype datatype that makes it extremely convenient to convert strings that are in standard date formates to datetime objects and then use those datetime objects to either create new features on a dataframe or work with the dataset in a timeseries fashion. 

This section will demonstrate how to take a column of date strings, convert it to a datetime object and then use the datetime formatting `.dt` to access specific parts of the date (year, month, day) to generate useful columns on a dataframe.

## Follow Along

### Work with Dates 

pandas documentation
- [to_datetime](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html)
- [Time/Date Components](https://pandas.pydata.org/pandas-docs/stable/timeseries.html#time-date-components) "You can access these properties via the `.dt` accessor"

Many of the most useful date columns in this dataset have the suffix `_d` to indicate that they correspond to dates.

We'll use a list comprehension to print them out

In [86]:
lstats['issue_d']

0         Dec-2018
1         Dec-2018
2         Dec-2018
3         Dec-2018
4         Dec-2018
            ...   
128407    Oct-2018
128408    Oct-2018
128409    Oct-2018
128410    Oct-2018
128411    Oct-2018
Name: issue_d, Length: 128412, dtype: object

Lets look at the string format of the `issue_d` column

In [87]:
[col for col in lstats if col.endswith('_d')]

['issue_d', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d']

Because this string format %m-%y is a common datetime format, we can just let Pandas detect this format and translate it to the appropriate datetime object.

In [89]:
lstats['issue_d'] = pd.to_datetime(lstats['issue_d'], infer_datetime_format=True)

Now we can see that the `issue_d` column has been changed to hold `datetime` objects.

Lets look at one of the cells specifically to see what a datetime object looks like:

In [90]:
lstats['issue_d'][0]

Timestamp('2018-12-01 00:00:00')

You can see how the month and year have been indicated by the strings that were contained in the column previously, and that the rest of the values have been inferred.

We can use the `.dt` accessor to now grab specific parts of the datetime object. Lets grab just the year from the all of the cells in the `issue_d` column

In [91]:
lstats['issue_d'].dt.year

0         2018
1         2018
2         2018
3         2018
4         2018
          ... 
128407    2018
128408    2018
128409    2018
128410    2018
128411    2018
Name: issue_d, Length: 128412, dtype: int64

Now the month.

In [92]:
lstats['issue_d'].dt.month

0         12
1         12
2         12
3         12
4         12
          ..
128407    10
128408    10
128409    10
128410    10
128411    10
Name: issue_d, Length: 128412, dtype: int64

It's just that easy! Now, instead of printing them out, lets add these year and month values as new columns on our dataframe. Again, you'll have to scroll all the way over to the right in the table to see the new columns.

In [93]:
lstats['issue_d_year'] = lstats['issue_d'].dt.year
lstats['issue_d_month'] = lstats['issue_d'].dt.month
lstats.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0      27975        27975          27975.0   36 months     14.47       962.52   
1      10000        10000          10000.0   60 months     12.98       227.43   
2      28000        28000          28000.0   60 months     13.56       645.15   
3      16000        16000          16000.0   60 months     13.56       368.66   
4       7500         7500           7500.0   36 months     10.72       244.55   

  grade sub_grade        emp_title emp_length home_ownership  annual_inc  \
0     C        C2        Conductor  10+ years       MORTGAGE    180000.0   
1     B        B5          Printer    9 years           RENT     60000.0   
2     C        C1  Project Manager  10+ years       MORTGAGE    128500.0   
3     C        C1             Lmsw  10+ years           RENT     46000.0   
4     B        B2  Program Manager    2 years           RENT     84000.0   

  verification_status    issue_d  loan_status pymnt_plan             purpose  \
0        Not Verified 2018-12-01   Fully Paid          n         credit_card   
1        Not Verified 2018-12-01      Current          n         credit_card   
2     Source Verified 2018-12-01      Current          n         credit_card   
3        Not Verified 2018-12-01      Current          n         credit_card   
4        Not Verified 2018-12-01  Charged Off          n  debt_consolidation   

                     title zip_code addr_state    dti  delinq_2yrs  \
0  Credit card refinancing    117xx         NY  11.47            0   
1  Credit card refinancing    660xx         KS  14.90            0   
2  Credit card refinancing    760xx         TX  27.35            1   
3  Credit card refinancing    787xx         TX  11.09            1   
4       Debt consolidation    600xx         IL   3.86            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0         Jul-1995               0                    39.0   
1         May-2007               0                     NaN   
2         Jul-1987               0                    24.0   
3         Oct-1997               1                    11.0   
4         May-2005               1                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN        10        0      29711      66.8%         19   
1                   112.0         7        1      10677      54.2%         12   
2                     NaN        16        0      55206        75%         31   
3                     NaN         9        0      18946      32.1%         21   
4                   114.0         9        1       2200      12.6%         17   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00  31804.529849   
1                   w    7352.95        7352.95   4487.310000   
2                   w   20667.91       20667.91  12871.360000   
3                   w   11503.84       11503.84   7570.880000   
4                   w       0.00           0.00   2829.930000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0         31804.53         27975.00        3829.53                 0.0   
1          4487.31          2647.05        1840.26                 0.0   
2         12871.36          7332.09        5539.27                 0.0   
3          7570.88          4496.16        3074.72                 0.0   
4          2829.93          1465.61         484.09                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0        0.00                   0.0000     Jan-2020         20288.02   
1        0.00                   0.0000     Aug-2020           227.43   
2        0.00                   0.0000     Aug-2020           645.15   
3        0.00                   0.0000     Aug-2020           368.66   
4      880.23                 158.4414     Sep-2019           244.55   

  next_pymnt_d last_credit_p

Because all of these dates come from Q4 of 2018, the `issue_d` column isn't all that interesting. Lets look at the `earliest_cr_line` column, which is also a string, but that could be converted to datetime format.

We're going to create a new column called `days_from_earliest_credit_to_issue`

It's a long column header, but think about how valuable this piece of information could be. This number will essentially indicate the length of a person's credit history and if that is correlated with repayment or other factors could be a valuable predictor!

In [95]:
lstats['earliest_cr_line'] = pd.to_datetime(lstats['earliest_cr_line'], infer_datetime_format=True)

What we're about to do is so cool! Pandas' datetime format is so smart that we can simply use the subtraction operator `-` in order to calculate the amount of time between two dates. 

Think about everything that's going on under the hood in order to give us such straightforward syntax! Handling months of different lengths, leap years, etc. Pandas datetime objects are seriously powerful!

In [98]:
lstats['days_from_earliest_credit_to_issue'] = (lstats['issue_d'] - lstats['earliest_cr_line']).dt.days
lstats.head()

loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment  \
0      27975        27975          27975.0   36 months     14.47       962.52   
1      10000        10000          10000.0   60 months     12.98       227.43   
2      28000        28000          28000.0   60 months     13.56       645.15   
3      16000        16000          16000.0   60 months     13.56       368.66   
4       7500         7500           7500.0   36 months     10.72       244.55   

  grade sub_grade        emp_title emp_length home_ownership  annual_inc  \
0     C        C2        Conductor  10+ years       MORTGAGE    180000.0   
1     B        B5          Printer    9 years           RENT     60000.0   
2     C        C1  Project Manager  10+ years       MORTGAGE    128500.0   
3     C        C1             Lmsw  10+ years           RENT     46000.0   
4     B        B2  Program Manager    2 years           RENT     84000.0   

  verification_status    issue_d  loan_status pymnt_plan             purpose  \
0        Not Verified 2018-12-01   Fully Paid          n         credit_card   
1        Not Verified 2018-12-01      Current          n         credit_card   
2     Source Verified 2018-12-01      Current          n         credit_card   
3        Not Verified 2018-12-01      Current          n         credit_card   
4        Not Verified 2018-12-01  Charged Off          n  debt_consolidation   

                     title zip_code addr_state    dti  delinq_2yrs  \
0  Credit card refinancing    117xx         NY  11.47            0   
1  Credit card refinancing    660xx         KS  14.90            0   
2  Credit card refinancing    760xx         TX  27.35            1   
3  Credit card refinancing    787xx         TX  11.09            1   
4       Debt consolidation    600xx         IL   3.86            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0       1995-07-01               0                    39.0   
1       2007-05-01               0                     NaN   
2       1987-07-01               0                    24.0   
3       1997-10-01               1                    11.0   
4       2005-05-01               1                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN        10        0      29711      66.8%         19   
1                   112.0         7        1      10677      54.2%         12   
2                     NaN        16        0      55206        75%         31   
3                     NaN         9        0      18946      32.1%         21   
4                   114.0         9        1       2200      12.6%         17   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00  31804.529849   
1                   w    7352.95        7352.95   4487.310000   
2                   w   20667.91       20667.91  12871.360000   
3                   w   11503.84       11503.84   7570.880000   
4                   w       0.00           0.00   2829.930000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0         31804.53         27975.00        3829.53                 0.0   
1          4487.31          2647.05        1840.26                 0.0   
2         12871.36          7332.09        5539.27                 0.0   
3          7570.88          4496.16        3074.72                 0.0   
4          2829.93          1465.61         484.09                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0        0.00                   0.0000     Jan-2020         20288.02   
1        0.00                   0.0000     Aug-2020           227.43   
2        0.00                   0.0000     Aug-2020           645.15   
3        0.00                   0.0000     Aug-2020           368.66   
4      880.23                 158.4414     Sep-2019           244.55   

  next_pymnt_d last_credit_p

In [99]:
lstats['years_from_earliest_credit_to_issue'] = lstats['days_from_earliest_credit_to_issue'] / 365
lstats['years_from_earliest_credit_to_issue'].describe()

count    128412.000000
mean         16.054497
std           7.908317
min           3.084932
25%          11.093151
50%          14.427397
75%          19.846575
max          68.961644
Name: years_from_earliest_credit_to_issue, dtype: float64

What's oldest credit history that was involved in Q4 2018? 

25,171 days is ~ 68.96 years of credit history!

## Challenge

Pandas' datetime format is so easy to work with that there's really no excuse for not using dates to make features on a dataframe! Get ready to practice more of this on your assignment.